crates/cairo1-run/src/lib.rs

```
@@ -284,9 +285,16 @@ pub fn run_program_at_path(filename: &PathBuf) -> Result<RunResult, Error> {
     let air_private_input: Option<PathBuf> = None;
     let memory_file: Option<PathBuf> = None;
 
+    // configure diagnostics
+    let mut program_diagnostics: Vec<String> = Vec::new();
+    let diagnostics_callback = |severity, diagnostic| {
+        program_diagnostics.push(format!("{severity}: {diagnostic}"));
+    };
+    let diagnostics_reporter = DiagnosticsReporter::callback(diagnostics_callback).allow_warnings();
+
     let compiler_config = CompilerConfig {
         replace_ids: true,
-        diagnostics_reporter: DiagnosticsReporter::default().allow_warnings(),
+        diagnostics_reporter,
         ..CompilerConfig::default()
     };
     let sierra_program = compile_cairo_project_at_path(filename, compiler_config)
```

Diagnostics can return both warnings and errors.  I noticed there is an early return here, when the cairo project fails to compile to sierra.

Doesn't this swallow all errors meaning only warnings will get to the frontend?

Rather than just returning a status code on failure (EXPECTATION_FAILED) we could also put a json body holding the diagnostics error. I think I'd just do that in this PR.

---

You are right, in the current implementation, errors don't reach the frontend. Let's do as you suggested.